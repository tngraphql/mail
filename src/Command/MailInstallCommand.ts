/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 4/8/2020
 * Time: 6:41 AM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import { BaseCommand, flags } from '@tngraphql/console';
import { Filesystem } from '@poppinss/dev-utils/build';
import { Application } from '@tngraphql/illuminate';
import { KernelContract } from '@tngraphql/console/dist/Contracts';
import * as path from 'path';

export class MailInstallCommand extends BaseCommand {
    static commandName = 'install:mail';

    static description = 'Copy layout mail views file';

    @flags.boolean({ description: 'Install the file even if the file already exists' })
    public force: boolean;

    private fs = new Filesystem(this.app.basePath('resources/views'));

    constructor(protected app: Application, kernel: KernelContract) {
        super(app, kernel);
    }

    async handle(...args: any[]): Promise<any> {
        const files = [
            'button.edge',
            'footer.edge',
            'header.edge',
            'layout.edge',
            'message.edge',
            'order.edge',
            'subcopy.edge'
        ];

        for( let file of files ) {
            const exists = await this.alreadyExists(file);

            if ( exists && ! this['force'] ) {
                this.logger.skip(`${ file } already exists`);
                return false;
            }
        }

        await Promise.all(files.map(fileName => {
            return this.copyfile(fileName);
        }));

        return undefined;
    }

    private async copyfile(name: string) {
        await this.fs.add(name, await this.fs.get(path.join(__dirname, '../resources/views', name)));
        this.logger.create(path.join(this.fs.basePath, name));
    }

    private async alreadyExists(file: string): Promise<boolean> {
        return this.fs.exists(file);
    }
}
