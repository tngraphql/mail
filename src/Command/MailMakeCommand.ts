/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 4/1/2020
 * Time: 10:03 AM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import { GeneratorCommand } from '@tngraphql/illuminate/dist/Foundation';
import * as path from "path";

export class MailMakeCommand extends GeneratorCommand {
    protected getStub(): string {
        return path.join(__dirname, 'stub/mail.stub');
    }

    protected getSuffix(): string {
        return '';
    }

    static commandName = 'make:mail';

    static description = 'Create a new email class';

    handle(...args: any[]): Promise<any> {
        return super.handle(...args);
    }

    getDestinationPath() {
        return path.join(this.application.getBasePath(), 'app/Mail');
    }

}
