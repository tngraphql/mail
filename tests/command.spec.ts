/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 4/6/2020
 * Time: 7:30 PM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import { Application, ConsoleKernel } from '@tngraphql/illuminate';
import { MailMakeCommand } from '../src/Command/MailMakeCommand';
import { join } from "path";
import { Filesystem } from '@poppinss/dev-utils/build';
import * as path from 'path';
import { MailInstallCommand } from '../src/Command/MailInstallCommand';

function toNewlineArray (contents: string): string[] {
    return contents.split(/\r?\n/)
}

const fs = new Filesystem(join(__dirname, 'app'));
describe('Command', () => {
    describe('Make | Mail', () => {
        afterEach(async () => {
            await fs.cleanup();
        });

        it('should create file mail.', async () => {
            const app = new Application(fs.basePath);
            app.environment = 'test';
            const kernel = new ConsoleKernel(app);
            const makeCommand = new MailMakeCommand(app, kernel.getAce());
            makeCommand.name = 'OrderCreate';

            await makeCommand.handle();

            expect(makeCommand.logger.logs[0].startsWith('underline(green(create))')).toBeTruthy();

            const schemaTemplate = await fs.get((makeCommand as any).getStub());
            const file = await fs.get(join(makeCommand.getDestinationPath(), 'OrderCreate.ts'));
            expect((makeCommand as any).getSuffix()).toBe('');

            expect(toNewlineArray(file)).toEqual(toNewlineArray(
                schemaTemplate
                    .replace('${filename}', 'OrderCreate')
                    .replace('${filename}', 'OrderCreate'),
            ));

            expect(file).toMatchSnapshot();
        });

        it('should skip when file exists', async () => {
            const app = new Application(fs.basePath);
            app.environment = 'test';
            const kernel = new ConsoleKernel(app);
            const makeCommand = new MailMakeCommand(app, kernel.getAce());
            makeCommand.name = 'OrderCreate';

            await makeCommand.handle();
            await makeCommand.handle();

            expect(makeCommand.logger.logs[1].startsWith('underline(magenta(skip))')).toBeTruthy();
        });

        it('should create file class email when file exists using force', async () => {
            const app = new Application(fs.basePath);
            app.environment = 'test';
            const kernel = new ConsoleKernel(app);
            const makeCommand = new MailMakeCommand(app, kernel.getAce());
            makeCommand.name = 'OrderCreate';

            await makeCommand.handle();

            const reMakeCommand = new MailMakeCommand(app, kernel.getAce());
            reMakeCommand.name = 'OrderCreate';
            reMakeCommand.force = true;
            await reMakeCommand.handle();

            expect(reMakeCommand.logger.logs[0].startsWith('underline(green(create))')).toBeTruthy();
        });
    });

    describe('Install | Mail', () => {
        afterEach(async () => {
            await fs.cleanup();
        });

        it('should install views mail', async () => {
            const app = new Application(fs.basePath);
            app.environment = 'test';
            const kernel = new ConsoleKernel(app);
            const install = new MailInstallCommand(app, kernel.getAce());

            await install.handle();

            expect(await fs.get('resources/views/button.edge')).toMatchSnapshot();
            expect(await fs.get('resources/views/footer.edge')).toMatchSnapshot();
            expect(await fs.get('resources/views/header.edge')).toMatchSnapshot();
            expect(await fs.get('resources/views/layout.edge')).toMatchSnapshot();
            expect(await fs.get('resources/views/message.edge')).toMatchSnapshot();
            expect(await fs.get('resources/views/order.edge')).toMatchSnapshot();
            expect(await fs.get('resources/views/subcopy.edge')).toMatchSnapshot();
        });

        it('should skip when file exists', async () => {
            const app = new Application(fs.basePath);
            app.environment = 'test';
            const kernel = new ConsoleKernel(app);
            const install = new MailInstallCommand(app, kernel.getAce());

            await install.handle();

            const reinstall = new MailInstallCommand(app, kernel.getAce());

            await reinstall.handle();
            expect(reinstall.logger.logs).toHaveLength(1);
            expect(reinstall.logger.logs[0].startsWith('underline(magenta(skip))')).toBeTruthy();
        });

        it('should install views mail when file exists using force', async () => {
            const app = new Application(fs.basePath);
            app.environment = 'test';
            const kernel = new ConsoleKernel(app);
            const install = new MailInstallCommand(app, kernel.getAce());

            await install.handle();

            const reinstall = new MailInstallCommand(app, kernel.getAce());
            reinstall.force=true;

            await reinstall.handle();
            expect(reinstall.logger.logs).toHaveLength(7);
            expect(reinstall.logger.logs[0].startsWith('underline(green(create))')).toBeTruthy();
        });
    });
})
