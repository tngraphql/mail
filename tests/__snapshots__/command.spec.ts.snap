// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command make command should work properly 1`] = `
"/**
 * (c) Phan Trung NguyÃªn <nguyenpl117@gmail.com>
 * User: nguyenpl117
 * Date: 4/6/2020
 * Time: 7:38 PM
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import { Mailable } from '@tngraphql/mail';

export class OrderCreate extends Mailable {

    /**
     * Create a new message instance.
     */
    public constructor() {
        super();
    }

    /**
     * Build the message.
     */
    public build(): void {
        //
    }
}
"
`;
